### DISCUSSION TITLE
Automatic re-render when sorting array

### DISCUSSION THREAD

**Discussion by PhilippWeiss85** [neuefische/questions#313] [Web] [answered]
Hi!
I have a small problem  that I am struggling with quite a bit: 
I want to add sort functions to my gamelist page where all the gamecards are displayed. A small menu appears on click on the icon in the top right corner and in this menu one should be able to click on another button to sort the array of the initial list and re-render it.
However, my page only reloads after closing the toggleMenu or directly onClick in the toggleMenu.

Related PR: https://github.com/PhilippWeiss85/capstone-project/pull/42

**Comment by ThomasF85** [comment, marked as answer]
Hi Philipp,

your problem occurs because you have committed a grave mortal sin of react development: mutating state
```
function sortGamesByName() {
    gameList.sort((a, b) => ...);
  }
```
You are taking the state `gameList` and changing (mutating) it directly. React does not rerender if a state object `gameList` changes. However, react realizes when the state `gameList` is a different object than before (and then triggers a rerender). You should always create a new modified object and then set the state to the new object:
```
function sortGamesByName() {
    const newGameList = ... // create a copy of the gamelist and sort it here
    setGameList(newGameList);
  }
```

**Reply by ThomasF85** [reply]
I just realized that you are using zustand. In that case you should give your zustand store a function that sorts the gamelist:
```
sortGames: () => {
      set((state) => {
        const sortedGames = ...
        return { games: sortedGames };
      });
    },
```
You should never just take a state variable and manipulate it directly. Always use functions that you define in the store (and that use the `set` you get from zustands `create((set) => ...)`) to change your state.

**Reply by PhilippWeiss85** [reply]
Hi @ThomasF85,
I just updated my code and transfered the sort functions to the useStore. However the re-rendering of the list still behaves the same. I still have to close the toggleMenu in order to get the gamecard list updated. 

My workaround for now is just to call my toggleMenu function within the sort functions.. so that when I call a sort function the togglemenu disapears and the list gets re-rendered. 
Is there still something wrong in my code? (updated the PR)


  ```sortGamesByName: () => {
      const toggleSort = get().nameToggle;
      const sortedGamesByName = ....
      get().toggleSortMenu();        // toggle the sortMenu to refresh the list
      set({
        games: sortedGamesByName,
        nameToggle: !toggleSort,
      });
    },

**Reply by ThomasF85** [reply]
The problem is that you still manipulate the same array and `sortedGamesByName` is a reference to the same array as before, but this very same array is now sorted. The Array.prototype.sort() function sorts the array in place and returns a reference to the same array. If you write it like this instead it should work:
```
const sortedGamesByName = [...get().games].sort((a, b) => ...)
```

**Reply by PhilippWeiss85** [reply]
Thank you so much Thomas! Works perfectly.
